pkg/executor/mpp_gather.go

```diff
@@ -104,19 +161,32 @@ func (e *MPPGather) Open(ctx context.Context) (err error) {
 		_, e.kvRanges, err = plannercore.GenerateRootMPPTasks(e.Ctx(), e.startTS, e.gatherID, e.mppQueryID, sender, e.is)
 		return err
 	}
-	planIDs := collectPlanIDS(e.originalPlan, nil)
-	e.gatherID = allocMPPGatherID(e.Ctx())
-	coord := e.buildCoordinator(planIDs)
-	err = mppcoordmanager.InstanceMPPCoordinatorManager.Register(mppcoordmanager.CoordinatorUniqueID{MPPQueryID: e.mppQueryID, GatherID: e.gatherID}, coord)
-	if err != nil {
+	if err = e.setupRespIter(ctx, false); err != nil {
 		return err
 	}
-	var resp kv.Response
-	resp, e.kvRanges, err = coord.Execute(ctx)
-	if err != nil {
-		return errors.Trace(err)
+
+	holdCap := mathutil.Max(32, mppErrRecoveryHoldChkCap*e.Ctx().GetSessionVars().MaxChunkSize)
+
+	disaggTiFlashWithAutoScaler := config.GetGlobalConfig().DisaggregatedTiFlash && config.GetGlobalConfig().UseAutoScaler
+	// For now, mpp err recovery only support MemLimit, which is only useful when AutoScaler is used.
+	enableMPPRecovery := disaggTiFlashWithAutoScaler
+
+	failpoint.Inject("mpp_recovery_test_force_enable", func() {
+		enableMPPRecovery = true
+	})
+
+	// No need to recovery when can fallback to tikv.
+	_, allowTiFlashFallback := e.Ctx().GetSessionVars().AllowFallbackToTiKV[kv.TiFlash]
+	if allowTiFlashFallback {
+		enableMPPRecovery = false
```

Comment:
Seems it needs to be moved before L174, otherwise it will conflict with `mpp_recovery_test_force_enable`.
why not just
```
// For now, mpp err recovery only support MemLimit, which is only useful when AutoScaler is used.
// No need to recovery when can fallback to tikv.
enableMPPRecovery := disaggTiFlashWithAutoScaler && !allowTiFlashFallback
```

Reply:
ok. I use different if branch to make the logic cleaner.

Reply:
`mpp_recovery_test_force_enable` is only for case, and `allowTiFlashFallback` is default to be off.

Reply:
and we can add test to ensure when `allowTiFlashFallback` is true, mpp err recovery should not enable.
and maybe rename `mpp_recovery_test_force_enable` as `mpp_recovery_test_mock_enable` is better.

Reply:
done